benchmarks/Lua/som.lua

```diff
@@ -0,0 +1,631 @@
+       current = current.next
+    end
+
+    if lo_tail then
+        lo_tail.next = nil
+        self.buckets[i] = lo_head
+    end
+    if hi_tail then
+        hi_tail.next = nil
+        self.buckets[i + old_storage.n] = hi_head
+    end
+end
+
+function Dictionary:remove_all ()
+    self.buckets = {n = self.buckets.n}
+    self.size = 0
+end
+
+function Dictionary:keys ()
+    local keys = Vector.new(self.size)
+    local buckets = self.buckets
+    for i = 1, buckets.n do
+        local current = buckets[i]
+        while current do
+            keys:append(current.key)
+            current = current.next
+        end
+    end
+    return keys
+end
+
+function Dictionary:values ()
+    local vals = Vector.new(self.size)
+    local buckets = self.buckets
+    for i = 1, buckets.n do
+        local current = buckets[i]
+        while current do
+            vals:append(current.value)
+            current = current.next
+        end
+    end
+    return vals
+end
+
+end -- class Dictionary
+
+local IdEntry = {_CLASS = 'IdEntry'} do
+setmetatable(IdEntry, {__index = Entry})
+
+function IdEntry.new (hash, key, value, next)
+    local obj = Entry.new (hash, key, value, next)
+    return setmetatable(obj, {__index = IdEntry})
+end
+
+function IdEntry:match (hash, key)
+    return self.hash == hash and self.key == key
+end
+
+end -- class IdEntry
+
+local IdentityDictionary = {_CLASS = 'IdentityDictionary'} do
+setmetatable(IdentityDictionary, {__index = Dictionary})
+
+function IdentityDictionary.new (size)
+    local obj = Dictionary.new (size)
+    return setmetatable(obj, {__index = Dictionary})
+end
+
+function IdentityDictionary:new_entry (key, value, hash)
+    return IdEntry.new(hash, key, value, nil)
+end
+
+end -- class IdentityDictionary
+
+local LuaDictionary = {_CLASS = 'LuaDictionary'} do
+
+function LuaDictionary.new ()
+    local obj = {
+        size = 0,
+        dict = {},
+    }
+    return setmetatable(obj, {__index = LuaDictionary})
+end
+
+function LuaDictionary:is_empty ()
+    return self.size == 0
+end
+
+function LuaDictionary:at (key)
+    return self.dict[key]
+end
+
+function LuaDictionary:at_put (key, value)
+    local current = self.dict[key]
+    if not current then
+        self.size = self.size + 1
+    end
+    self.dict[key] = value
+end
+
+function LuaDictionary:remove_all ()
+    self.dict = {}
+    self.size = 0
+end
+
+function LuaDictionary:keys ()
+    local keys = Vector.new(self.size)
+    for k in pairs(self.dict) do
+        keys:append(k)
+    end
+    return keys
+end
+
+function LuaDictionary:values ()
+    local values = Vector.new(self.size)
+    for _, v in pairs(self.dict) do
+        values:append(v)
+    end
+    return values
+end
+
+end -- class LuaDictionary
+
+local Random = {_CLASS = 'Random'} do
+
+function Random.new ()
+    local obj = {seed = 74755}
+    return setmetatable(obj, {__index = Random})
+end
+
+function Random:next ()
+  self.seed = ((self.seed * 1309) + 13849) % 65536;
```

Comment:
refactored, see https://github.com/smarr/are-we-fast-yet/pull/30/commits/a981c7f062ecdbac641824cc9f421341a9b808f7